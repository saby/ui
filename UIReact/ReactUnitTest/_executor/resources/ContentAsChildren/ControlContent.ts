import { Control, TWasabyOverReactProps } from 'UICore/Base';
import template = require('wml!ReactUnitTest/_executor/resources/ContentAsChildren/ControlContent');

interface IControlContentProps extends TWasabyOverReactProps {
    allowPatchContent: boolean;
}

export default class ControlContent extends Control<IControlContentProps> {
    protected _template = template;
}
