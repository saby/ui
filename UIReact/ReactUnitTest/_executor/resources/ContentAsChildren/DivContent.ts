import { Control, TWasabyOverReactProps } from 'UICore/Base';
import template = require('wml!ReactUnitTest/_executor/resources/ContentAsChildren/DivContent');

interface IDivContentProps extends TWasabyOverReactProps {
    allowPatchContent: boolean;
}

export default class DivContent extends Control<IDivContentProps> {
    protected _template = template;
}
