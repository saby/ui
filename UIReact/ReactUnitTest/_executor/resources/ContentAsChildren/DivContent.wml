<div>
   <ws:if data="{{ _options.allowPatchContent }}">
      <ReactUnitTest._executor.resources.ContentAsChildren.PatchChildrenInsert>
         <div class="div-content">Div content</div>
      </ReactUnitTest._executor.resources.ContentAsChildren.PatchChildrenInsert>
   </ws:if>
   <ws:else>
      <ReactUnitTest._executor.resources.ContentAsChildren.SimpleChildrenInsert>
         <div class="div-content">Div content</div>
      </ReactUnitTest._executor.resources.ContentAsChildren.SimpleChildrenInsert>
   </ws:else>
</div>
