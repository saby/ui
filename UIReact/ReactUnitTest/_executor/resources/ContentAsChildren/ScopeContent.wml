<div>
   <ReactUnitTest._executor.resources.ContentAsChildren.SomeWasabyWithContent>
      <ws:content>
         <div class="objectContentInScope">
            <ws:if data="{{ _options.allowPatchContent }}">
               <ReactUnitTest._executor.resources.ContentAsChildren.PatchChildrenInsert scope="{{ _options }}" />
            </ws:if>
            <ws:else>
               <ReactUnitTest._executor.resources.ContentAsChildren.SimpleChildrenInsert scope="{{ _options }}" />
            </ws:else>
         </div>
      </ws:content>
   </ReactUnitTest._executor.resources.ContentAsChildren.SomeWasabyWithContent>
</div>
