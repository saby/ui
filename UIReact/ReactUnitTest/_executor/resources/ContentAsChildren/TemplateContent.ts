import { Control, TWasabyOverReactProps } from 'UICore/Base';
import template = require('wml!ReactUnitTest/_executor/resources/ContentAsChildren/TemplateContent');

interface ITemplateContentProps extends TWasabyOverReactProps {
    allowPatchContent: boolean;
}

export default class TemplateContent extends Control<ITemplateContentProps> {
    protected _template = template;
}
