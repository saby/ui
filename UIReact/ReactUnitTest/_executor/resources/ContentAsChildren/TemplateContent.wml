<ws:template name="inline">
   <div class="template-content">Template content</div>
</ws:template>

<div>
   <ws:if data="{{ _options.allowPatchContent }}">
      <ReactUnitTest._executor.resources.ContentAsChildren.PatchChildrenInsert>
         <ws:partial template="inline" />
      </ReactUnitTest._executor.resources.ContentAsChildren.PatchChildrenInsert>
   </ws:if>
   <ws:else>
      <ReactUnitTest._executor.resources.ContentAsChildren.SimpleChildrenInsert>
         <ws:partial template="inline" />
      </ReactUnitTest._executor.resources.ContentAsChildren.SimpleChildrenInsert>
   </ws:else>
</div>
