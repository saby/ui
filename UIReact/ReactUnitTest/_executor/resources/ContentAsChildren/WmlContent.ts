import { Control, TWasabyOverReactProps } from 'UICore/Base';
import template = require('wml!ReactUnitTest/_executor/resources/ContentAsChildren/WmlContent');

interface IWmlContentProps extends TWasabyOverReactProps {
    allowPatchContent: boolean;
}

export default class WmlContent extends Control<IWmlContentProps> {
    protected _template = template;
}
