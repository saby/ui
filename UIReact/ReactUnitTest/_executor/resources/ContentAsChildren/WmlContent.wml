<div>
   <ws:if data="{{ _options.allowPatchContent }}">
      <ReactUnitTest._executor.resources.ContentAsChildren.PatchChildrenInsert>
         <ws:partial template="wml!ReactUnitTest/_executor/resources/ContentAsChildren/Wml" />
      </ReactUnitTest._executor.resources.ContentAsChildren.PatchChildrenInsert>
   </ws:if>
   <ws:else>
      <ReactUnitTest._executor.resources.ContentAsChildren.SimpleChildrenInsert>
         <ws:partial template="wml!ReactUnitTest/_executor/resources/ContentAsChildren/Wml" />
      </ReactUnitTest._executor.resources.ContentAsChildren.SimpleChildrenInsert>
   </ws:else>
</div>
