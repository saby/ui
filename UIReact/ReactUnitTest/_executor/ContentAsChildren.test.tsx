/**
 * @jest-environment jsdom
 */
import { render, unmountComponentAtNode } from 'react-dom';
import { act } from 'react-dom/test-utils';

import ControlContent from './resources/ContentAsChildren/ControlContent';
import DivContent from './resources/ContentAsChildren/DivContent';
import WmlContent from './resources/ContentAsChildren/WmlContent';
import TemplateContent from './resources/ContentAsChildren/TemplateContent';

// Именно шаблон без контрола, чтобы scope="{{ _options }}" работал как старый scope="{{ ... }}"
import ScopeContent = require('wml!ReactUnitTest/_executor/resources/ContentAsChildren/ScopeContent');
import CloneContentAsChildrenParent from './resources/ContentAsChildren/CloneContentAsChildrenParent';

describe('ContentAsChildren', () => {
    let container: HTMLDivElement;

    beforeEach(() => {
        container = document.createElement('div', {});
        document.body.appendChild(container);
    });
    afterEach(() => {
        unmountComponentAtNode(container);
        container.remove();
        container = null;
    });

    describe.each([
        ['Реакт компонент рисует полученный children неизменным', false],
        ['Реакт компонент мутирует полученный children', true],
    ])('%s', (describeName, allowPatchContent) => {
        it.each([
            ['Контент - wasaby контрол', ControlContent],
            ['Контент - div элемент', DivContent],
            ['Контент - wml шаблон', WmlContent],
            ['Контент - inline шаблон', TemplateContent],
            ['Контент - объект из скоупа, не нужно превращать в Children', ScopeContent],
        ])('%s', (name, RootComponent) => {
            act(() => {
                render(<RootComponent allowPatchContent={allowPatchContent} />, container);
            });
            expect(container).toMatchSnapshot();
        });
    });

    it('Пропсы чистого реакта, заданные через cloneElement ContentAsChildren, не летят ниже корня контента', () => {
        act(() => {
            render(<CloneContentAsChildrenParent />, container);
        });

        expect(container).toMatchSnapshot('реф, класс и data-атрибут не пролетели жальше корня');
    });
});
