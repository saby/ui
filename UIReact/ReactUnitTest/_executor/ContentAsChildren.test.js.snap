// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ContentAsChildren Реакт компонент мутирует полученный children Контент - div элемент 1`] = `
<div>
  <div
    tabindex="0"
  >
    <div
      class="patch"
    >
      <div
        class="div-content patched"
        tabindex="0"
      >
        Div content
      </div>
    </div>
  </div>
</div>
`;

exports[`ContentAsChildren Реакт компонент мутирует полученный children Контент - inline шаблон 1`] = `
<div>
  <div
    tabindex="0"
  >
    <div
      class="patch"
    >
      <div
        class="template-content patched"
        tabindex="0"
      >
        Template content
      </div>
    </div>
  </div>
</div>
`;

exports[`ContentAsChildren Реакт компонент мутирует полученный children Контент - wasaby контрол 1`] = `
<div>
  <div
    tabindex="0"
  >
    <div
      class="patch"
    >
      <div
        class="control-content patched"
        tabindex="0"
      >
        Control content
      </div>
    </div>
  </div>
</div>
`;

exports[`ContentAsChildren Реакт компонент мутирует полученный children Контент - wml шаблон 1`] = `
<div>
  <div
    tabindex="0"
  >
    <div
      class="patch"
    >
      <div
        class="wml-content patched"
        tabindex="0"
      >
        Wml content
      </div>
    </div>
  </div>
</div>
`;

exports[`ContentAsChildren Реакт компонент мутирует полученный children Контент - объект из скоупа, не нужно превращать в Children 1`] = `
<div>
  <div
    tabindex="0"
  >
    <div
      class="paintContentPatchOptions"
      tabindex="0"
    >
      <div
        class="objectContentInScope"
        tabindex="0"
      >
        <div
          class="patch classNameFromScope"
        >
          No children
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`ContentAsChildren Реакт компонент рисует полученный children неизменным Контент - div элемент 1`] = `
<div>
  <div
    tabindex="0"
  >
    <div
      class="simple"
    >
      <div
        class="div-content"
        tabindex="0"
      >
        Div content
      </div>
    </div>
  </div>
</div>
`;

exports[`ContentAsChildren Реакт компонент рисует полученный children неизменным Контент - inline шаблон 1`] = `
<div>
  <div
    tabindex="0"
  >
    <div
      class="simple"
    >
      <div
        class="template-content"
        tabindex="0"
      >
        Template content
      </div>
    </div>
  </div>
</div>
`;

exports[`ContentAsChildren Реакт компонент рисует полученный children неизменным Контент - wasaby контрол 1`] = `
<div>
  <div
    tabindex="0"
  >
    <div
      class="simple"
    >
      <div
        class="control-content"
        tabindex="0"
      >
        Control content
      </div>
    </div>
  </div>
</div>
`;

exports[`ContentAsChildren Реакт компонент рисует полученный children неизменным Контент - wml шаблон 1`] = `
<div>
  <div
    tabindex="0"
  >
    <div
      class="simple"
    >
      <div
        class="wml-content"
        tabindex="0"
      >
        Wml content
      </div>
    </div>
  </div>
</div>
`;

exports[`ContentAsChildren Реакт компонент рисует полученный children неизменным Контент - объект из скоупа, не нужно превращать в Children 1`] = `
<div>
  <div
    tabindex="0"
  >
    <div
      class="paintContentPatchOptions"
      tabindex="0"
    >
      <div
        class="objectContentInScope"
        tabindex="0"
      >
        <div
          class="simple classNameFromScope"
        >
          No children
        </div>
      </div>
    </div>
  </div>
</div>
`;
